{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Building and analyzing enhancer-driven TF-gene regulatory networks PANDA and LIONESS\n",
    "\n",
    "Author: Romana T. Pop<sup>1</sup>, Maud Fagny<sup>2</sup>\n",
    "\n",
    "<sup>1</sup>Centre for Molecular Medicine Norway (NCMM), University of Oslo, Oslo, Norway\n",
    "\n",
    "<sup>2</sup>Université Paris-Saclay, INRAE, CNRS, AgroParisTech, GQE – Le Moulon, Gif-sur-Yvette, France\n",
    "\n",
    "# Introduction\n",
    "Enhancers are key regulators of gene activity in Eukaryotes, especially during development. Enhancers regulates their target gene by binding transcription factors (TFs) that interact with target gene promoters through 3D chromatin loops<sup>1</sup>. The role of enhancers in wiring the gene regulatory networks is less understood in plants than in animals<sup>2</sup>. Recently, genome-wide studies have allowed to identify many putative active enhancers in plants<sup>3, 4</sup>. The distance between enhancers and their nearest genes was found to vary between species, largely based on genome size<sup>5</sup>. In maize, studies of 3D chromatin data (HiChIP) suggest that 25% of enhancers skip the nearest gene and regulate one further away, and that at least 34% of them regulate multiple genes<sup>4,6</sup>. Finally, it has been shown that genes involved in the same biological pathways tend to be regulated by enhancers carrying binding sites for similar transcription factors<sup>4</sup>, thus creating genome-wide gene regulatory networks that allow the activation of many genes from the same process by using a small number of transcription factors. Identifying enhancers' target genes and deciphering the genome-wide gene regulatory network they articulate is thus an important challenge. It is also still unclear how enhancers and their genome-wide gene regulatory networks arise in plants. It has been proposed that a transposable elements (TEs), which are mobile elements that can duplicate and insert themselves at different loci in the genome, carrying their own regulatory elements, may play a role<sup>7</sup>. \n",
    "\n",
    "This tutorial illustrates the analysis that was performed by Fagny et al.<sup>8</sup> on maize data from Oka et al.<sup>7</sup>. Here, we compare the regulatory networks of two types of maize leaves at different developmental stages: dividing young leaves in seedlings (V2-IST) and mature leaves protecting the cob (Husk). We look at differential regulation between the two tissues, identify common and tissue specific regulatory network modules and investigate the role of TEs in wiring stage-specific networks.   \n",
    "\n",
    "# Necessary data and software\n",
    "\n",
    "If you are running this locally or adapting it to other data, before beginning the analysis, make sure to install all the necessary software and properly pre-process the data.\n",
    "\n",
    "This analysis can be run on the server or locally by setting the runserver parameter to 0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "runserver=1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also need to set the paths for data in the server."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if (runserver==1){\n",
    "    ppath='/opt/data/netZooR/maizeenh/'\n",
    "} else if (runsevrer==0) {\n",
    "    ppath=''\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In addition, some sections of the case study can require some time to run, therefore, we set the precomputed parameter to 1 to load precomputed data. Set to 0 to run the whole tutorial from scratch."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "precomputed=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "if(runserver==0){\n",
    "    install.packages(\"devtools\")\n",
    "    library(devtools)\n",
    "    install.packages(\"purr\")\n",
    "    install.packages(\"tidyr\")\n",
    "    install.packages(\"data.table\")\n",
    "    install.packages(\"RColorBrewer\")\n",
    "    install.packages(\"umap\")\n",
    "    install.packages(\"ggplot2\")\n",
    "    install.packages(\"igraph\")\n",
    "    BiocManager::install(\"IRanges\")\n",
    "    BiocManager::install(\"limma\")\n",
    "    BiocManager::install(\"topGO\")\n",
    "    BiocManager::install(\"Rgraphviz\")\n",
    "    BiocManager::install(\"GOstats\")\n",
    "    devtools::install_github(\"netZoo/netZooR\", build_vignettes = TRUE)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The pipeline requires the following data as input:\n",
    "\n",
    "* ***lioness_output.Rdata*** : an Rdata file containing the lioness networks for all samples (as a list of data frames);\n",
    "* ***panda_output.Rdata*** : an Rdata file containing the PANDA output (as a data frame);\n",
    "* ***samples_annotation.Rdata*** : an Rdata file containing the sample annotation (as a data frame);\n",
    "* ***gene_expression.Rdata***: an Rdata file containing the gene expression data used to compute the panda networks (as a data frame, optional)\n",
    "* ***GeneOntologyAnnotation.txt*** : an Rdata file containing the gene ontology annotation (as a txt file);\n",
    "* ***enhancers_genes.Rdata*** : an Rdata file containing each enhancer-gene candidate pair (as a data frame);\n",
    "* ***enhancers_TE_overlaps.Rdata*** : an Rdata file containing information about the colocalization of TEs and enhancers (as a data frame);\n",
    "* ***tfbs_enhancers_TE_genes.Rdata*** : an Rdata file containing information about the colocalization of TEs and TFBS, and the link between enhancers, their predicted TFBS and their potential target genes (as a data frame);\n",
    "* ***corres_TF_athaliana_zmays.Rdata*** : an Rdata file containing a match table between the names of the transcription factors in *Arabidopsis thaliana* and their orthologs in *Zea mays mays*.\n",
    "\n",
    "In this tutorial, we will use the enhancer data from Oka et al<sup>7</sup>: a list of enhancers active in either V2-IST or Husk. We integrate these data with gene expression data from the same tissues (6 replicates/tissues). To obtain more reliable gene co-expression data, we also integrate expression data from 11 other tissues from the same maize line obtained in another experiment (AMAIZING GeneAtlas<sup>8</sup>). We will not focus here on generating regulatory networks with PANDA/LIONESS, as this is covered in other tutorials:\n",
    "\n",
    "[Building PANDA and LIONESS Regulatory Networks from GTEx Gene Expression Data in R](http://netbooks.networkmedicine.org/user/c14c992d-0826-4d80-9f82-c303d37a5167/notebooks/netZooR/ApplicationinGTExData.ipynb)\n",
    "\n",
    "[Comparing state-specific PANDA networks using pandaR](http://netbooks.networkmedicine.org/user/c14c992d-0826-4d80-9f82-c303d37a5167/notebooks/netZooR/pandaR.ipynb)\n",
    "\n",
    "[Generating 26 cancer gene regulatory network using TCGA datasets](http://netbooks.networkmedicine.org/user/99920d65-89aa-401b-a223-7dc30bf999f2/notebooks/netZooR/tcga_networks.ipynb)\n",
    "\n",
    "## A note on priors when inferring enhancers-derived gene regulatory networks\n",
    "\n",
    "One notable difference between the panda priors used in the tutorials above and the prior used for this analysis is linked to the inference of gene regulatory networks articulated by enhancers instead of promoters. Unlike promoters, enhancers may regulate more than one gene, have a much wider region of action and their nearest gene may not always be their target, particularly in large genomes. Therefore, a slightly different approach was used to create the PANDA prior for this analysis. Enhancers were first annotated with TF binding sites (TFBSs) using FIMO<sup>9</sup>. Because enhancers regulate genes at a distance and can regulate genes located both upstream and downstream, we selected 250kb windows on either side of each enhancer and scanned for genes in these windows. Finally, to account for the fact that enhancers may not always regulate their nearest genes or may regulate multiple genes, we created an edge between each TF that had a predicted binding site in an enhancer and all the genes within +/- 250kb of that enhancer. \n",
    "\n",
    "Prior workflow:\n",
    "\n",
    "* Load expression, TF binding site (TFBS) and enhancer data;\n",
    "* Extract significant TFBSs predicted in enhancers;\n",
    "* Select 250kb windows either side of the enhancers;\n",
    "* Find genes in the +/- 250kb windows on either side of the enhancer (potential target genes);\n",
    "* Create edges between TF who has a predicted binding site in an enhancer and each potential target gene.\n",
    "\n",
    "# Load data and libraries\n",
    "\n",
    "Now that we have installed all the software and prepared the necessary data, we can begin the analysis. First we load the libraries and the data.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "#loading libraries\n",
    "library(umap)\n",
    "library(RColorBrewer)\n",
    "library(limma)\n",
    "library(topGO)\n",
    "library(Rgraphviz)\n",
    "library(netZooR)\n",
    "library(topGO)\n",
    "library(data.table)\n",
    "library(tidyr)\n",
    "library(purrr)\n",
    "library(IRanges)\n",
    "library(ggplot2)\n",
    "library(igraph)\n",
    "\n",
    "#loading the panda and lioness networks & GO annotation\n",
    "load(paste0(ppath, \"lioness_output.Rdata\"))\n",
    "load(paste0(ppath, \"panda_output.Rdata\"))\n",
    "load(paste0(ppath, \"samples_annotation.Rdata\"))\n",
    "load(paste0(ppath, \"gene_expression.Rdata\"))\n",
    "GO.file <- paste0(ppath, \"GeneOntologyAnnotation.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Differential targeting analysis between Husk and V2-IST tissues\n",
    "\n",
    "In this section, we will look at differential targeting between Husk and V2-IST tissues by comparing our PANDA/LIONESS networks. First, we perform a log transformation on the LIONESS network edge weights, to ensure that none of the logtransformed edge weight has a negative or infinite value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "#function that performs the log transformation\n",
    "logtrans <- function(x) {\n",
    "  y <- log(exp(x)+1)\n",
    "  y[y==Inf] <- x[y==Inf] # replace any infinite values with the original edge values # nolint\n",
    "  return(y)\n",
    "}\n",
    "\n",
    "#performing the transformation on our data\n",
    "if(precomputed==0){\n",
    "    ind_net_log <- lapply(lioness, logtrans)\n",
    "} else {\n",
    "    load(paste0(ppath, \"logtransformed_lioness_results.Rdata\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The PANDA algorithm requires a prior network as input. All potential TF-gene regulatory relationships are integrated in the prior, and edge weights are set to 1. The edge weights are then iteratively updated by PANDA. New edges can also be learnt by integrating information from protein-protein interactions and gene co-expression data. Thus the PANDA and LIONESS networks have more edges than the prior network. Here, we specifically focus on interactions between predicted TFBSs in enhancers and their target genes. Although potential non-canonical or indirect interactions could be revealed through analysis of non-prior edges, we decided to specifically focus on direct interactions, and thus we filter out non-prior edges from the lioness networks. We also filter out any edges that do not take at least 2 different values among the sample-specific networks built from the six replicates of either Husk or V2-IST. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if(precomputed==0){\n",
    "    #filtering out all the additional edges yielded by PANDA/LIONESS and keeping only prior edges # nolint\n",
    "    ind_net_log_tab <- NULL\n",
    "    for(i in 1:length(lioness)) {\n",
    "      ind_net_log_tab <- cbind(ind_net_log_tab, as.numeric(t(as.matrix(ind_net_log[[i]]))[panda$prior == 1])) # filtering non-prior edges in the log-transformed networks # nolint\n",
    "    }\n",
    "\n",
    "    #reformatting\n",
    "    ind_net_log_tab <- data.frame(t(ind_net_log_tab))\n",
    "    colnames(ind_net_log_tab) <- paste(panda$Gene, panda$TF, sep=\"_\")[panda$prior == 1] # nolint\n",
    "    rownames(ind_net_log_tab) <- names(ind_net_log)\n",
    "\n",
    "    #removing edges that do not have different values in at least 2 samples from Husk or V2-IST \n",
    "    #necessary for differential targeting analysis\n",
    "    tissues <- samples.annot[rownames(ind_net_log_tab), \"tissue\"]\n",
    "    ind_net_log_tab[,apply(ind_net_log_tab, 2, function(x) length(unique(x))) <= 2] <- NULL # nolint\n",
    "} else {\n",
    "    load(paste0(ppath, \"logtransformed_lioness_results_table.Rdata\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Although PANDA relies on expression data to generate the networks, PANDA edges and expression data cluster differently from one another. To show this, we plot a UMAP of the gene expression levels and the network edge weights (this step is optional and not necessary for the rest of the analysis). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#umap visualisation of edges and expression\n",
    "custom_config <- umap.defaults\n",
    "if(nrow(ind_net_log_tab) <= 15) {\n",
    "  custom_config$n_neighbors <- nrow(ind_net_log_tab) - 1\n",
    "}\n",
    "custom_config$n_neighbors <- 6\n",
    "custom_config$random_state <- 1 # Set the seed\n",
    "expr_df <- as.data.frame(t(expr)) # expression data\n",
    "expr_umap <- umap(expr_df[,apply(expr_df, 2, max) <= 2], custom_config) # nolint\n",
    "ind_net_log_umap <- umap(ind_net_log_tab[,apply(ind_net_log_tab, 2, max) <= 2], custom_config) # nolint\n",
    "\n",
    "#adding a column for tissue colours to the sample annotation\n",
    "#tissue_colour <- palette.colors(n = length(unique(samples.annot$tissue)), palette = \"Polychrome 36\") # nolint\n",
    "tissue_colour <- c('#5A5156','#E4E1E3','#F6222E','#FE00FA','#16FF32','#3283FE','#FEAF16','#B00068','#1CFFCE','#90AD1C','#2ED9FF','#DEA0FD','#AA0DFE')\n",
    "names(tissue_colour) <- unique(samples.annot$tissue)\n",
    "smpl_split <- split(samples.annot, samples.annot$tissue)\n",
    "\n",
    "for(i in names(smpl_split)) {\n",
    "  smpl_split[[`i`]]$tissue_colour <- tissue_colour[i]\n",
    "}\n",
    "\n",
    "samples.annot <- unsplit(smpl_split, samples.annot$tissue)\n",
    "\n",
    "#plot umap of gene expression and network edges\n",
    "layout(matrix(c(1,2), 1, 2, byrow = TRUE), respect = T, heights = c(20,20), widths = c(20,20))\n",
    "par(mar = c(3, 3, 5, 1) + .1, xpd = TRUE)\n",
    "plot(expr_umap$layout[, 1], expr_umap$layout[, 2],\n",
    "     col = samples.annot[rownames(expr_umap$layout), \"tissue_colour\"],\n",
    "     pch = ifelse(samples.annot[rownames(expr_umap$layout), \"tissue\"] %in% c(\"Husk\", \"V2-IST\"), # nolint\n",
    "     17, 16), xlab = \"\", ylab = \"\", cex.lab = 1, cex.axis = 1)\n",
    "legend(par(\"usr\")[1]-(par(\"usr\")[2]-par(\"usr\")[1])*1/4, par(\"usr\")[4]+(par(\"usr\")[4]-par(\"usr\")[3])*1/2, \n",
    "       legend = names(tissue_colour), col = tissue_colour,\n",
    "       pch = ifelse(names(tissue_colour) %in% c(\"Husk\", \"V2-IST\"), 17, 16), ncol = 4, bty = 'n')  # nolint\n",
    "par(mar = c(3, 3, 5, 1) + .1, xpd = TRUE)\n",
    "plot(ind_net_log_umap$layout[, 1], ind_net_log_umap$layout[, 2],\n",
    "     col = samples.annot[rownames(ind_net_log_umap$layout), \"tissue_colour\"],\n",
    "     pch = ifelse(samples.annot[rownames(ind_net_log_umap$layout), \"tissue\"] %in% c(\"Husk\", \"V2-IST\"), # nolint\n",
    "     17, 16), xlab = \"\", ylab = \"\", cex.lab = 1, cex.axis = 1)\n",
    "legend(par(\"usr\")[1]-(par(\"usr\")[2]-par(\"usr\")[1])*1/4, par(\"usr\")[4]+(par(\"usr\")[4]-par(\"usr\")[3])*1/2, \n",
    "       legend = names(tissue_colour), col = tissue_colour,\n",
    "       pch = ifelse(names(tissue_colour) %in% c(\"Husk\", \"V2-IST\"), 17, 16), ncol = 4, bty = 'n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we use limma to perform a differential targeting analysis between Husk and V2-IST tissues. First, we must reformat the data for use with limma and chose a *p*-value significance threshold for the differential analysis. We can then proceed to run the differential targeting analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#limma\n",
    "#set threshold for p-values\n",
    "thres <- 0.01\n",
    "\n",
    "#prepare data for limma\n",
    "tissues <- samples.annot[rownames(ind_net_log_tab), \"tissue\"]\n",
    "limmadiffexpr <- list()\n",
    "limmadiffexprsignif <- list()\n",
    "tab <- ind_net_log_tab[tissues %in% c(\"Husk\", \"V2-IST\"),]\n",
    "tissues_list <- tissues[tissues %in% c(\"Husk\", \"V2-IST\")]\n",
    "\n",
    "#create a binary vector with 0 representing HUSK samples and 1 representing V2-IST samples # nolint\n",
    "tissue <- rep(0, nrow(tab))\n",
    "tissue[tissues_list == \"V2-IST\"] <- 1\n",
    "\n",
    "# Run limma analysis\n",
    "design <- model.matrix(~ tissue)\n",
    "fit <- lmFit(t(tab), design = design)\n",
    "fit <- eBayes(fit)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We extract the significant results for Husk and V2-IST, and print the number of differentially targeted genes and the number of TFs targeting these genes. We see that there are 2123 genes that are more targeted in Husk than V2-IST by 55 TFs, and 2075 genes that are more targeted by 54 TFs in V2-IST than in Husk. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract results\n",
    "limmadiffexpr[[\"V2\"]] <- topTable(fit, number = ncol(tab))\n",
    "limmadiffexpr[[\"husk\"]] <- topTable(fit, number = ncol(tab))\n",
    "\n",
    "# extract results below the significance threshold\n",
    "limmadiffexprsignif[[\"V2\"]] <- limmadiffexpr[[\"V2\"]][limmadiffexpr[[\"V2\"]]$adj.P.Val <= thres & #nolint\n",
    "                                    limmadiffexpr[[\"V2\"]]$logFC > 0, ]\n",
    "limmadiffexprsignif[[\"husk\"]] <- limmadiffexpr[[\"husk\"]][limmadiffexpr[[\"husk\"]]$adj.P.Val <= thres & #nolint\n",
    "                                    limmadiffexpr[[\"husk\"]]$logFC < 0, ]\n",
    "                                            \n",
    "#extracting differentially targeted genes\n",
    "a <- lapply(limmadiffexprsignif, rownames)\n",
    "diff_genes <- lapply(a, function(x){unique(unlist(lapply(x, function(y) strsplit(y, \"_\")[[1]][1])))})\n",
    "nb_diff_genes <- unlist(lapply(diff_genes, function(x) length(unique(x)))) #number of differentially targeted genes in each tissue\n",
    "print(\"Differentially targeted genes:\")\n",
    "nb_diff_genes\n",
    "\n",
    "#extracting differentially targeting TFs\n",
    "diff_TFs <- lapply(a, function(x){unlist(lapply(x, function(y) strsplit(y, \"_\")[[1]][2]))})\n",
    "nb_diff_TFs <- unlist(lapply(diff_TFs, function(x) length(unique(x)))) #number of differentially targeted genes in each tissue\n",
    "print(\"Differentially targeting TFs:\")\n",
    "nb_diff_TFs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After identifying the differentially targeted genes, we can look for functional enrichment among them. To do this, we will perform a Gene Ontology (GO) enrichment analysis. To account for the nested structure of the Gene Ontology, we used the elim algorithm from the R Bioconductor topGO package, and the Fisher’s exact test to test for enrichment significance. Because we will perform GO enrichment analysis again later on, we write a function that performs it. The function takes as input the following:\n",
    "* ***geneNames*** - gene names from the GO annotation\n",
    "* ***list.topgenes*** - differentially expressed genes\n",
    "* ***all.genes*** - list of genes in our data\n",
    "* ***geneID2GO*** - the mapping of the GO genes to the GO annotation\n",
    "* ***max.nb.genes*** - the maximum number of genes to show\n",
    "* ***tag*** - the tissue of interest\n",
    "* ***plots*** - logical; indicates whether to save plots of enriched GO terms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#function for GO\n",
    "GOenrich <- function(geneNames, list.topgenes, all.genes, geneID2GO, max.nb.genes, tag, plots = FALSE) {\n",
    "  allRes <- list()\n",
    "  allRes[[tag]] <- list()\n",
    "  GOtermsGenes <- list()\n",
    "  GOtermsGenes[[tag]] <- list()\n",
    "\n",
    "  for(i in names(list.topgenes)) {\n",
    "    allRes[[tag]][[i]] <- list()\n",
    "    if(length(list.topgenes[[i]]) >= 5) {\n",
    "      myInterestingGenes <- unique(list.topgenes[[i]]) #nolint\n",
    "      geneList <- factor(as.integer(geneNames[geneNames %in% all.genes] %in% myInterestingGenes)) #nolint\n",
    "      names(geneList) <- geneNames[geneNames %in% all.genes]\n",
    "      GOtermsGenes[[tag]][[i]] <- data.frame(\"Ontology\" = character(0),\n",
    "                                              \"GOTerm\" = character(0),\n",
    "                                              \"Genes\" = character(0))\n",
    "      for(ont in c(\"BP\", \"MF\")) {\n",
    "        GOdata <- new(\"topGOdata\", ontology = ont, allGenes = geneList,\n",
    "                    annot = annFUN.gene2GO, gene2GO = geneID2GO, nodeSize = 5)\n",
    "        ## Compute stats\n",
    "        test.stat <- new(\"elimCount\", testStatistic = GOFisherTest, name = \"Fisher test\")#nolint\n",
    "        resultElimFisher <- getSigGroups(GOdata, test.stat)\n",
    "        allRes[[tag]][[i]][[`ont`]] <- GenTable(GOdata, elim = resultElimFisher, orderBy = \"elim\", #nolint\n",
    "                                           ranksOf = \"elim\", topNodes = max.nb.genes) #nolint     \n",
    "        allRes[[tag]][[i]][[`ont`]]$ontology <- rep(ont, nrow(allRes[[tag]][[i]][[`ont`]]))#nolint\n",
    "        \n",
    "        #plot\n",
    "        if(plots == TRUE) {\n",
    "          pdf(file = paste0(\"TopGONodes_\", ont, \".pdf\"))\n",
    "          showSigOfNodes(GOdata, score(resultElimFisher), firstSigNodes = 10, useInfo = 'all')#nolint\n",
    "          dev.off()\n",
    "        }\n",
    "        \n",
    "        \n",
    "        myterms <- allRes[[tag]][[i]][[`ont`]]$GO.ID\n",
    "        mygenes  <- genesInTerm(GOdata, myterms)\n",
    "        \n",
    "        for (j in 1:length(myterms)) {\n",
    "          myterm <- myterms[j]\n",
    "          mygenesforterm <- mygenes[myterm][[1]]\n",
    "          myfactor <- mygenesforterm %in% myInterestingGenes # find the genes that are in the list of genes of interest #nolint\n",
    "          mygenesforterm2 <- mygenesforterm[myfactor == TRUE]\n",
    "          mygenesforterm2 <- paste(mygenesforterm2, collapse = \",\")\n",
    "          GOtermsGenes[[tag]][[i]] <- rbind(GOtermsGenes[[tag]][[i]],\n",
    "                                    data.frame(\"Ontology\" = ont, \"GOTerm\" = myterm, \"Genes\" = mygenesforterm2))#nolint\n",
    "        }\n",
    "      }\n",
    "      allRes[[tag]][[i]]=rbind(allRes[[tag]][[i]]$BP, allRes[[tag]][[i]]$MF)\n",
    "    }\n",
    "  }\n",
    "  return(list(allRes, GOtermsGenes))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have this function, we can perform the GO enrichment analysis on differentially targeted genes from each tissue. We sort the results by *p*-values and we see that among the most significant enriched terms for Husk, some are related to leaf growth and differentiation (cell growth, vacuole organization, root hair cell differentiation), protein activity and post translational protein modifications (protein serine/threonine kinase activity, protein dimerization activity, protein glycosylation, protein targeting to vacuole), and signal transduction (magnesium ion transport, signal transduction). For the V2-IST tissue, we find an enrichment in processes related to cell division (cytokinesis by cell plate formation,  regulation of G2/M transition of mitotic cell cycle, spindle assembly, cell population proliferation), leaf morphogenesis and photosynthesis (carotenoid biosynthetic process, photosynthesis, chlorophyll binding, photosynthetic electron transport chain) and gene expression regulation (gene silencing). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if(precomputed==0){\n",
    "    #prepare GO annotation\n",
    "    all_genes=unique(unlist(lapply(colnames(ind_net_log_tab), function(x){strsplit(x, \"_\")[[1]][1]})))\n",
    "    ZmB73_5a_xref.GO.topGO.FG <- read.delim(GO.file, header=FALSE, stringsAsFactors = F)\n",
    "    geneNames<-ZmB73_5a_xref.GO.topGO.FG$V1\n",
    "    geneID2GO <- readMappings(file = GO.file)\n",
    "   \n",
    "    #perform GO enrichment analysis\n",
    "    diff_GO_res <- GOenrich(geneNames = geneNames, geneID2GO = geneID2GO,\n",
    "                        list.topgenes = diff_genes, all.genes = all_genes,\n",
    "                        tag = \"V2\", max.nb.genes = 100)\n",
    "} else {\n",
    "    load(paste0(ppath, \"GO_diff_targeted_genes.Rdata\"))\n",
    "}\n",
    "\n",
    "#extract tissue specific enrichment\n",
    "husk <- diff_GO_res[[1]][[1]]$husk\n",
    "V2 <- diff_GO_res[[1]][[1]]$V2\n",
    "\n",
    "#order enriched terms by most significant hits\n",
    "husk <- husk[order(husk$elim), ]\n",
    "V2 <- V2[order(V2$elim), ]\n",
    "cat(\"Husk:\\n\")\n",
    "head(husk, n = 15)\n",
    "cat(\"V2-IST:\\n\")\n",
    "head(V2, n = 15)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identification of top enhancer targets\n",
    "\n",
    "We can also identify the top target gene for each enhancer. This is done as follows: for each enhancer-gene pair, we compute an average targeting score. To this end, we extract all of the TFs with predicted binding sites in an enhancer. We then compute the average of the edge weight between each TF and one target gene. The gene with the highest average targeting score is the most likely target.\n",
    "\n",
    "*  This section needs as input:\n",
    "    + A data frame with samples annotations (samples.annot);\n",
    "    + A data frame with the lioness log-transformed edges (ind_net_log_tab)\n",
    "    + A data frame linking the predicted TFBS, the enhancers, and their potential target genes (enh.genes).\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(paste0(ppath, \"enhancers_genes.Rdata\"))\n",
    "tis <- samples.annot$tissue \n",
    "\n",
    "enh <- sort(unique(c(enh.genes$enh))) # Extract list of enhancers\n",
    "edges <- t(ind_net_log_tab[rownames(samples.annot),])\n",
    "edges <- as.data.frame(edges[,tis %in% c(\"Husk\", \"V2-IST\")]) # Create edges table\n",
    "\n",
    "### Functions\n",
    "## For each enhancer-gene pair, compute average TF-gene edge value across all TF with predicted binding site in the enhancer\n",
    "average.edge <- function(geneID, tfbs, edges, gene) {#nolint\n",
    "    tfs <- tfbs$Alias[tfbs$geneID == geneID]\n",
    "    res <- NULL\n",
    "    for(tf in tfs) {\n",
    "        alltfs <- unlist(strsplit(tf, \";\"))\n",
    "        e <- edges[paste(geneID, alltfs, sep = \"_\"), ]\n",
    "        res <- rbind(res, colSums(e, na.rm = T) / nrow(e))\n",
    "    }\n",
    "    if(is.null(res)) { return(rep(0,12)) } else { return(colSums(res) / nrow(res)) }#nolint\n",
    "}\n",
    "\n",
    "## Find enhancer-gene pairs and compute average edge for each of them\n",
    "compute.average.edge <- function(enhID, pairs, edges) {#nolint\n",
    "  tmp <- pairs[pairs$enh == enhID, ]\n",
    "  g <- unique(tmp$geneID)\n",
    "  if(length(g) > 0 & !all(is.na(g))) {\n",
    "    score.genes <- t(apply(data.frame(g, stringsAsFactors = F), 1, average.edge, \n",
    "                           tfbs = tmp, edges = edges))\n",
    "        rownames(score.genes) <- g\n",
    "    if(nrow(score.genes) == 1) {\n",
    "      avg.husk <- mean(score.genes[, 1:6])\n",
    "      avg.V2 <- mean(score.genes[, 7:12])\n",
    "      names(avg.husk) <- names(avg.V2) <- g\n",
    "    } else {\n",
    "      avg.husk=rowSums(score.genes[,1:6])/ncol(score.genes[,1:6])\n",
    "      avg.V2=rowSums(score.genes[,7:12])/ncol(score.genes[,7:12])\n",
    "    }\n",
    "    out <- data.frame(\"enh\" = rep(enhID, length(avg.husk)), \"gene\" = names(avg.husk), #nolint\n",
    "                       \"edge.husk\" = avg.husk, \"top.husk\"=ifelse(avg.husk == max(avg.husk), 1, 0), #nolint\n",
    "                       \"edge.V2\" = avg.V2, \"top.V2\" = ifelse(avg.V2 == max(avg.V2), 1, 0),#nolint\n",
    "                       stringsAsFactors = F)\n",
    "  } else {\n",
    "    out <- data.frame(\"enh\" = enhID, \"gene\" = NA, \"edge.husk\" = NA, \"top.husk\" = NA,\n",
    "                     \"edge.V2\" = NA, \"top.V2\" = NA, stringsAsFactors = F)\n",
    "  }\n",
    "  return(out)\n",
    "}\n",
    "\n",
    "### Compute targeting for all enhancer-gene pairs and pick highest target\n",
    "if(precomputed==0){\n",
    "    enh.gene.pairs <- t(apply(data.frame(enh, stringsAsFactors=F), 1, compute.average.edge, \n",
    "                          pairs = enh.genes, edges = edges)) #nolint\n",
    "} else {\n",
    "    load(paste0(ppath, \"enhancer_gene_pairs.Rdata\"))\n",
    "}\n",
    "\n",
    "##Make result table\n",
    "res <- NULL\n",
    "for(i in 1:length(enh.gene.pairs)) {\n",
    "  res <- rbind(res, enh.gene.pairs[[i]])\n",
    "}\n",
    "rownames(res) <- NULL\n",
    "head(res[res$top.husk ==1 | res$top.V2==1,],15)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identification of tissue-specific modules with CONDOR & ALPACA\n",
    "\n",
    "To further characterize differential targeting between Husk and V2-IST, we will identify tissue-specific communities in our regulatory networks using CONDOR and ALPACA. First, we will compute the average LIONESS edges across six replicates for each tissue, as it is the required input for ALPACA.\n",
    "\n",
    "* This section needs as input:\n",
    "    + A data frame with the lioness log-transformed edges (ind_net_log_tab)\n",
    "    + The original panda and lioness outputs (panda, lioness)\n",
    "    + A table with information on transposable elements and enhancer overlaps, and the potential target genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get average edges by tissue\n",
    "tmp <- by(ind_net_log_tab, samples.annot[rownames(ind_net_log_tab), \"tissue\"],\n",
    "                                function(x){colSums(x)/nrow(x)})\n",
    "edges_by_tissue <- data.frame(matrix(unlist(tmp),\n",
    "                             ncol=length(unique(samples.annot[rownames(ind_net_log_tab), \"tissue\"])))) #nolint\n",
    "colnames(edges_by_tissue) <- names(tmp)\n",
    "rownames(edges_by_tissue) <- names(tmp[[1]])\n",
    "rm(tmp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can identify communities in each tissue-specific network with CONDOR. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(paste0(ppath, \"corres_TF_athaliana_zmays.Rdata\"))\n",
    "### Compute tissue-specific condor networks\n",
    "# prepare condor input - extract edges for TFs that are expressed in at least one of the tissues\n",
    "maize_TF <- ortho[order(ortho$TF),]\n",
    "expr_tfs <- unique(maize_TF$TF[maize_TF$zmays_eg_homolog_ensembl_gene %in% rownames(expr)])\n",
    "core_tab <- lapply(rownames(edges_by_tissue), function(x) { return(unlist(strsplit(x, \"_\")[[1]]))}) #nolint\n",
    "core_tab <- data.frame(matrix(unlist(core_tab), ncol=2, byrow=TRUE),\n",
    "                       stringsAsFactors = FALSE)\n",
    "colnames(core_tab) <- c(\"Gene\", \"TF\")\n",
    "core_tab <- core_tab[core_tab$TF %in% expr_tfs, ]\n",
    "\n",
    "# run condor\n",
    "condor_network <- list()\n",
    "for (i in colnames(edges_by_tissue)[colnames(edges_by_tissue) %in% c(\"Husk\", \"V2-IST\")]) {\n",
    "  cat(\"Finding community structure for tissue \", i, \"\\n\")\n",
    "  tab <- cbind(core_tab, \"edge\" = edges_by_tissue[paste(core_tab$Gene, core_tab$TF, sep = \"_\"), i]) #nolint\n",
    "  tab <- tab[tab$edge != Inf, ]\n",
    "  condor_network[[`i`]] <- createCondorObject(tab, return.gcc=T)\n",
    "  condor_network[[`i`]] <- condorCluster(condor_network[[`i`]], project = F) #nolint\n",
    "}\n",
    "\n",
    "###Stats on networks\n",
    "nb_com <- sapply(condor_network, function(x) nrow(x$Qcoms)) #number of communities #nolint\n",
    "modularity <- sapply(condor_network, function(x) max(x$modularity)) #modularity\n",
    "cat(\"\\nNumber of communities by network:\\n\")\n",
    "nb_com\n",
    "cat(\"\\nNetworks modularities:\\n\")\n",
    "modularity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we will use ALPACA to compare the network structures of the two tissues. We first need to make a slight modification to the main alpaca function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#modified ALPACA function\n",
    "alpaca <- function (net.table, file.stem, verbose = T) {\n",
    "  net.table[, 1] <- paste(net.table[, 1], \"A\", sep = \"_\")\n",
    "  net.table[, 2] <- paste(net.table[, 2], \"B\", sep = \"_\")\n",
    "  print(\"Detecting communities in control network...\")\n",
    "  ctrl.pos <- net.table[net.table[, 3] >= 0, 1:3]\n",
    "  ctrl.elist <- data.frame(red = ctrl.pos[, 2], blue = ctrl.pos[, \n",
    "                                                                1], weights = ctrl.pos[, 3])\n",
    "  ctrl.condor <- createCondorObject(ctrl.elist)\n",
    "  ctrl.condor <- condorCluster(ctrl.condor, project = F)\n",
    "  ctrl.memb <- c(ctrl.condor$red.memb[, 2], ctrl.condor$blue.memb[, \n",
    "                                                                  2])\n",
    "  names(ctrl.memb) <- c(as.character(ctrl.condor$red.memb[, \n",
    "                                                          1]), as.character(ctrl.condor$blue.memb[, 1]))\n",
    "  if (!(is.null(file.stem))) \n",
    "    write.table(ctrl.memb, paste(c(file.stem, \"_ALPACA_ctrl_memb.txt\"), \n",
    "                                 collapse = \"\"), row.names = T, col.names = F, quote = F, \n",
    "                sep = \"\\t\")\n",
    "  pos.table <- net.table[intersect(which(net.table[, 3] >= \n",
    "                                           0), which(net.table[, 4] >= 0)), ]\n",
    "  pos.graph <- graph.edgelist(as.matrix(pos.table[, 1:2]), \n",
    "                              directed = T)\n",
    "  if (length(setdiff(V(pos.graph)$name, names(ctrl.memb))) > \n",
    "      0) {\n",
    "    uncounted <- setdiff(V(pos.graph)$name, names(ctrl.memb))\n",
    "    unc.memb <- sample(1:max(ctrl.memb), length(uncounted), \n",
    "                       replace = T)\n",
    "    names(unc.memb) <- uncounted\n",
    "    ctrl.memb <- c(ctrl.memb, unc.memb)\n",
    "  }\n",
    "  print(\"Computing differential modularity matrix...\")\n",
    "  dwbm <- alpacaComputeDWBMmatMscale(pos.table, ctrl.memb[V(pos.graph)$name])\n",
    "  if (verbose) {\n",
    "    write.table(dwbm, paste(c(file.stem, \"_DWBM.txt\"), collapse = \"\"), \n",
    "                row.names = T, col.names = T, quote = F, sep = \"\\t\")\n",
    "    write.table(rownames(dwbm), paste(c(file.stem, \"_DWBM_rownames.txt\"), \n",
    "                                      collapse = \"\"), row.names = T, col.names = T, quote = F, \n",
    "                sep = \"\\t\")\n",
    "    write.table(colnames(dwbm), paste(c(file.stem, \"_DWBM_colnames.txt\"), \n",
    "                                      collapse = \"\"), row.names = T, col.names = T, quote = F, \n",
    "                sep = \"\\t\")\n",
    "  }\n",
    "  ntfs <- nrow(dwbm)\n",
    "  ngenes <- ncol(dwbm)\n",
    "  this.B <- array(0, dim = c(ntfs + ngenes, ntfs + ngenes))\n",
    "  this.B[1:ntfs, (ntfs + 1):(ntfs + ngenes)] <- dwbm\n",
    "  this.B[(ntfs + 1):(ntfs + ngenes), 1:ntfs] <- t(dwbm)\n",
    "  print(\"Computing differential modules...\")\n",
    "  louv.memb <- alpaca.genlouvain(this.B)\n",
    "  names(louv.memb) <- c(rownames(dwbm), colnames(dwbm))\n",
    "  print(\"Computing node scores...\")\n",
    "  louv.Ascores <- NULL\n",
    "  louv.Bscores <- NULL\n",
    "  for (i in 1:max(louv.memb)) {\n",
    "    this.comm <- names(louv.memb)[louv.memb == i]\n",
    "    this.tfs <- this.comm[grep(\"_A\", this.comm)]\n",
    "    this.genes <- this.comm[grep(\"_B\", this.comm)]\n",
    "    if ((length(this.tfs) > 1 ) & (length(this.genes) > 1)) {\n",
    "      tf.sums <- apply(dwbm[this.tfs, this.genes], 1, sum)\n",
    "      gene.sums <- apply(dwbm[this.tfs, this.genes], 2, \n",
    "                         sum)\n",
    "    }        else {\n",
    "      tf.sums <- sum(dwbm[this.tfs, this.genes])\n",
    "      gene.sums <- dwbm[this.tfs, this.genes]\n",
    "    }\n",
    "    this.denom <- sum(dwbm[this.tfs, this.genes])\n",
    "    louv.Ascores <- c(louv.Ascores, tf.sums/this.denom)\n",
    "    louv.Bscores <- c(louv.Bscores, gene.sums/this.denom)\n",
    "  }\n",
    "  louv.scores <- c(louv.Ascores, louv.Bscores)\n",
    "  if (!is.null(file.stem)) {\n",
    "    write.table(cbind(names(louv.memb), as.vector(louv.memb)), \n",
    "                paste(c(file.stem, \"_ALPACA_final_memb.txt\"), collapse = \"\"), \n",
    "                col.names = F, row.names = F, quote = F, sep = \"\\t\")\n",
    "    write.table(cbind(names(louv.scores), louv.scores), paste(c(file.stem, \n",
    "                                                                \"_ALPACA_scores.txt\"), collapse = \"\"), row.names = F, \n",
    "                col.names = F, quote = F, sep = \"\\t\")\n",
    "  }\n",
    "  list(louv.memb, louv.scores)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ALPACA allows us to compare two networks and find differences in their community structure. We can use the network structure we identified with CONDOR to explore differences in the regulatory networks of the Husk and V2-IST tissues. This will allow us to identify genes and TFs that contributed to changes in the modularity between the two tissues. ALPACA considers one network as a background \"control\"; to account for this, we run ALPACA two times, using in turn the Husk and V2-IST networks as the background and we extract the top genes for each tissue. This can take a few minutes.\n",
    "\n",
    "When setting V2-IST as reference, we find 75 modules in the husk network, including 9 with at least 10 genes and one TF.\n",
    "\n",
    "When setting Husk as reference, we find 70 modules in the V2-IST network, including 11 with at least 10 genes and one TF."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#set up alpaca input object\n",
    "core_tab_alpaca <- core_tab[,2:1]\n",
    "max_nb_genes <- 100\n",
    "TFs <- paste0(\"tf\", 1:length(unique(core_tab$TF)))\n",
    "names(TFs) <- unique(core_tab$TF)\n",
    "genes <- paste0(\"gene\", 1:length(unique(core_tab$Gene)))\n",
    "names(genes) <- unique(core_tab$Gene)\n",
    "\n",
    "core_tab_alpaca <- core_tab\n",
    "core_tab_alpaca$Gene <- genes[core_tab_alpaca$Gene]\n",
    "core_tab_alpaca$TF <- TFs[core_tab_alpaca$TF]\n",
    "\n",
    "#alpaca on husk - set V2-IST as reference and compute the differential modularity between husk and V2-IST. \n",
    "set.seed(1)\n",
    "V2_husk_edges <- data.frame(core_tab[, 2:1],\n",
    "                        \"V2\" = edges_by_tissue[paste(core_tab$Gene, core_tab$TF, sep = \"_\"), \"V2-IST\"], #nolint\n",
    "                        \"husk\" = edges_by_tissue[paste(core_tab$Gene, core_tab$TF, sep = \"_\"), \"Husk\"], stringsAsFactors = FALSE)#nolint\n",
    "if (precomputed==0){\n",
    "    V2_husk <- alpaca(V2_husk_edges, file.stem = paste0(\"alpaca.results.V2ref\"), verbose=TRUE)#nolint\n",
    "} else {\n",
    "    load(paste0(ppath, \"alpaca_results.Rdata\"))\n",
    "}\n",
    "\n",
    "#List of genes in each community of husk network\n",
    "husk_genes <- tapply(names(V2_husk[[1]]), V2_husk[[1]], function(x) { gsub(\"_B\", \"\", x[grep(\"_B\", x)])})#nolint\n",
    "#List of TFs in each community of husk network\n",
    "husk_TFs <- tapply(names(V2_husk[[1]]), V2_husk[[1]], function(x) { gsub(\"_A\", \"\", x[grep(\"_A\", x)])})#nolint\n",
    "\n",
    "#List of genes in each community of husk network, with their score reflecting how much of the differential modularity they explain \n",
    "husk_genes_scores <- lapply(husk_genes, function(x, scores) {\n",
    "  res <- scores[paste0(x, \"_B\")]\n",
    "  names(res) <- x\n",
    "  return(sort(res))\n",
    "}, scores = V2_husk[[2]])\n",
    "\n",
    "#List of TFs in each community of husk network, with their score reflecting how much of the differential modularity they explain \n",
    "husk_TFs_scores <- lapply(husk_TFs, function(x, scores) {\n",
    "  res <- scores[paste0(x, \"_A\")]\n",
    "  names(res) <- x\n",
    "  return(sort(res))\n",
    "}, scores = V2_husk[[2]])\n",
    "\n",
    "#List of 100 top genes\n",
    "husk_topgenes <- lapply(husk_genes_scores, function(x) {\n",
    "  n <- names(sort(x, decreasing = T))[1:max_nb_genes]\n",
    "  return(n[!is.na(n)])})\n",
    "names(husk_topgenes) <- paste0(\"Husk_\", names(husk_topgenes))\n",
    "\n",
    "#alpaca on V2-IST - set husk as reference and compute the differential modularity between husk and V2-IST. \n",
    "husk_V2_edges <- data.frame(core_tab[, 2:1],\n",
    "                        \"husk\" = edges_by_tissue[paste(core_tab$Gene, core_tab$TF, sep = \"_\"), \"Husk\"], #nolint\n",
    "                        \"V2\" = edges_by_tissue[paste(core_tab$Gene, core_tab$TF, sep = \"_\"), \"V2-IST\"], stringsAsFactors = FALSE)#nolint\n",
    "if (precomputed==0){\n",
    "    husk_V2 <- alpaca(husk_V2_edges, file.stem = \"alpaca.results.huskref\", verbose = TRUE)#nolint\n",
    "}\n",
    "\n",
    "#List of genes in each community of V2-IST network\n",
    "V2_genes <- tapply(names(husk_V2[[1]]), husk_V2[[1]], function(x) { gsub(\"_B\", \"\", x[grep(\"_B\", x)])})#nolint\n",
    "\n",
    "#List of TFs in each community of V2-IST network\n",
    "V2_TFs <- tapply(names(husk_V2[[1]]), husk_V2[[1]], function(x) { gsub(\"_A\", \"\", x[grep(\"_A\", x)])})#nolint\n",
    "\n",
    "#List of genes in each community of V2-IST network, with their score reflecting how much of the differential modularity they explain \n",
    "V2_genes_scores <- lapply(V2_genes, function(x, scores) {\n",
    "  res <- scores[paste0(x, \"_B\")]\n",
    "  names(res) <- x\n",
    "  return(sort(res))\n",
    "}, scores = husk_V2[[2]])\n",
    "\n",
    "#List of TFs in each community of V2-IST network, with their score reflecting how much of the differential modularity they explain \n",
    "V2_TFs_scores <- lapply(V2_TFs, function(x, scores) {\n",
    "  res <- scores[paste0(x, \"_A\")]\n",
    "  names(res)=x\n",
    "  return(sort(res))\n",
    "}, scores = husk_V2[[2]])\n",
    "\n",
    "#List of 100 top genes\n",
    "V2_topgenes <- lapply(V2_genes_scores, function(x) {\n",
    "  n <- names(sort(x, decreasing = T))[1:max_nb_genes]\n",
    "  return(n[!is.na(n)])})\n",
    "names(V2_topgenes) <- paste0(\"V2-IST_\", names(V2_topgenes))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To compare the gene content of the communities in the Husk and V2-IST networks, we calculate the jaccard index of each pair of communities. The best corresponding community is the one with the highest jaccard index. If the best jaccard index is lower than 0.4, we consider a community as tissue-specific, otherwise as shared.\n",
    "\n",
    "We identify 2 tissue-specific communities in the Husk network, and 3 in the V2-IST network. We also identify 7 shared communities. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#compute jaccard index\n",
    "jaccard <- NULL\n",
    "for(i in 1:length(V2_genes)) {\n",
    "  tmp <- c()\n",
    "  for(j in 1:length(husk_genes)) {\n",
    "    tmp <- c(tmp, (sum(V2_genes[[i]] %in% husk_genes[[j]]) / length(union(V2_genes[[i]], husk_genes[[j]]))))#nolint\n",
    "  }\n",
    "  jaccard <- cbind(jaccard, tmp)\n",
    "}\n",
    "rownames(jaccard) <- paste(\"Husk\", 1:nrow(jaccard), sep = \"_\")\n",
    "colnames(jaccard) <- paste(\"V2-IST\", 1:ncol(jaccard), sep = \"_\")\n",
    "\n",
    "#extract tissue-specific communities\n",
    "husk_jaccard <- apply(jaccard, 1, max)[unlist(lapply(husk_TFs, length)) >= 1 & unlist(lapply(husk_genes, length)) >= 10] #nolint\n",
    "husk_unique <- husk_jaccard[husk_jaccard<0.4]\n",
    "cat(\"Husk top jaccard index for each community:\\n\")\n",
    "husk_jaccard\n",
    "\n",
    "V2_jaccard <- apply(jaccard, 2, max)[unlist(lapply(V2_TFs, length)) >=1 & unlist(lapply(V2_genes, length)) >= 10]#nolint\n",
    "V2_unique <- V2_jaccard[V2_jaccard<0.4]\n",
    "cat(\"V2-IST top jaccard index for each community:\\n\")\n",
    "V2_jaccard\n",
    "\n",
    "## Extract list of topgenes \n",
    "find.unique.husk <-function(i, h = husk_genes_scores) {\n",
    "  names(head(sort(h[[i]], decreasing = T), max_nb_genes))\n",
    "}\n",
    "find.unique.V2 <- function(i, v = V2_genes_scores){\n",
    "    names(head(sort(v[[i]], decreasing = T), max_nb_genes))\n",
    "}\n",
    "\n",
    "husk_unique_topgenes <- subset(husk_topgenes, names(husk_topgenes) %in% names(husk_unique))\n",
    "\n",
    "V2_unique_topgenes <- subset(V2_topgenes, names(V2_topgenes) %in% names(V2_unique))\n",
    "\n",
    "shared_topgenes <- subset(husk_topgenes, (!(names(husk_topgenes) %in% names(husk_unique)) & (names(husk_topgenes) %in% names(husk_jaccard))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once again, we perform Gene Ontology categories enrichment analysis among the top genes of each tissue-specific community for Husk and V2-IST networks, and shared communities between husk and V2-IST.\n",
    "The first Husk-specific community is enriched for functions linked to cell wall organization or biogenesis. The second one is enriched for functions linked to rRNA-containing ribonucleoprotein complex binding and oligosaccharide biosynthetic process.\n",
    "\n",
    "The first V2-IST-specific community is enriched for several functions linked to response to hormones (regulation of jasmonic acid mediated signal; cytokinin-activated signaling pathway); regulation of transcription (regulation of nucleic acid-templated transcription; transcription corepressor activity), and amino-acid biosynthetic processes (sulfur compound biosynthetic process; aromatic amino acid family biosynthetic process). The second V2-IST community is enriched in histone modification (cellular macromolecule biosynthetic process, histone methyltransferase activity) and mitotic process (sister chromatid cohesion). Finally, the third V2-IST-specific community is enriched in leaf vascular tissue pattern formation and response to hypoxia processes. All these functions are coherent with a dividing, differentiating seedling tissues.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Compute Gene Ontology enrichment analyses for each tissue-specufic regulatory community.\n",
    "if(precomputed==0){\n",
    "    ## V2-specific\n",
    "    V2_GO_res <- GOenrich(geneNames = geneNames, geneID2GO = geneID2GO,\n",
    "         list.topgenes = V2_unique_topgenes, all.genes = unique(unlist(V2_genes)), tag = \"V2-specific_communities\",  max.nb.genes = 100)\n",
    "    ## Husk-specific\n",
    "    Husk_GO_res <- GOenrich(geneNames = geneNames, geneID2GO = geneID2GO,\n",
    "         list.topgenes = husk_unique_topgenes, all.genes = unique(unlist(husk_genes)), tag = \"Husk-specific_communities\",  max.nb.genes = 100)\n",
    "} else {\n",
    "    load(paste0(ppath, \"GO_tissue_specific_communities.Rdata\"))\n",
    "}\n",
    "\n",
    "#Order enriched terms by most significant hits\n",
    "cat(\"V2-IST specific communities:\\n\")\n",
    "for(i in 1:length(V2_GO_res[[1]][[1]])){\n",
    "    tmp <- V2_GO_res[[1]][[1]][[i]]\n",
    "    tmp <- tmp[order(tmp$elim), ]\n",
    "    cat(\"\\n\", names(V2_GO_res[[1]][[1]])[i], \":\\n\", sep=\"\")\n",
    "    print(head(tmp[tmp$elim<=0.01,], n = 15))\n",
    "}\n",
    "\n",
    "cat(\"Husk specific communities:\\n\")\n",
    "for(i in 1:length(Husk_GO_res[[1]][[1]])){\n",
    "    tmp <- Husk_GO_res[[1]][[1]][[i]]\n",
    "    tmp <- tmp[order(tmp$elim), ]\n",
    "    cat(\"\\n\", names(Husk_GO_res[[1]][[1]])[i], \":\\n\", sep=\"\")\n",
    "    print(head(tmp[tmp$elim<=0.01,], n = 15))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Enrichment analysis for transposable elements families and superfamilies among enhancers and TFBS\n",
    "\n",
    "We can further analyse the gene regulatory networks and investigate the role of TEs in the tissue-specific regulation of gene expression. To this end, we look for relative enrichment of TE superfamilies and orders in tissue-specific enhancers and/or TFBS, and analyse the functions of the genes they regulate. But first, we investigate the proportion of active enhancers that are overlapping at least one TE in Husk and V2-IST, and search for enrichment in specific order of TEs (LTR, LINE, TIR, MITE, or Helitrons). The TE annotation of B73 was generated using RepeatMasker and an updated library of TE can be found here: https://github.com/oushujun/EDTA/blob/master/database/maizeTE11122019.\n",
    "\n",
    "We find that each category of enhancers is enriched in TE from different orders, suggesting that different types of TE were domesticated to build regulatory networks, both shared between tissues and tissue-specific. Husk-specific activated enhancers are enriched in Miniature inverted-repeat transposable elements (MITEs) while V2-IST-specific activated enhancers are not enriched in any specific TE order. Finally, enhancers active in both tissues are enriched in Long Terminal Repeats (LTRs)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load and preprocess data\n",
    "load(paste0(ppath, \"enhancers_TE_overlaps.Rdata\"))\n",
    "enh.TE <- enh.TE[enh %in% enh.genes$enh]\n",
    "tissue.enh <- unique(enh.TE[, c(\"enh.pos\", \"tissue\")])$tissue\n",
    "names(tissue.enh) <- unique(enh.TE[, c(\"enh.pos\", \"tissue\")])$enh.pos\n",
    "\n",
    "### For each enhancer, determine if it is overlapping TE from zero (\"None\"), one or multiple orders\n",
    "res=unlist(by(enh.TE, enh.TE$enh.pos, function(annot){\n",
    "    annot.order=unique(annot$order)\n",
    "    if(length(annot.order[!(annot.order %in% \"None\")])>1){\n",
    "        annot.order <- \"Multiple\"\n",
    "    }\n",
    "    if(length(annot.order[!(annot.order %in% \"None\")])==1){\n",
    "        annot.order <- annot.order[!(annot.order %in% \"None\")]\n",
    "    }\n",
    "    return(annot.order)\n",
    "}))\n",
    "\n",
    "### For each category of enhancers, compute the proportion of enhancers overlapping TEs from each order\n",
    "colors.TE <- c(\"white\", \"gold\", \"forestgreen\", \"green\", \"blue\", \"cyan\", \"purple\")\n",
    "names(colors.TE) <- c(\"None\", \"Multiple\", \"LTR\", \"LINE\", \"TIR\", \"MITE\", \"Helitron\")\n",
    "husk.enh.TE <- summary(factor(res[(tissue.enh[names(res)] %in% \"Husk\")], level=unique(as.character(res))))\n",
    "V2.enh.TE <- summary(factor(res[(tissue.enh[names(res)] %in% \"V2-IST\")], level=unique(as.character(res))))\n",
    "Both.enh.TE <- summary(factor(res[(tissue.enh[names(res)] %in% \"Both\")], level=unique(as.character(res))))\n",
    "All.enh.TE <- summary(factor(res, level=unique(as.character(res))))\n",
    "prop=cbind(\"All\"=All.enh.TE[names(colors.TE)]/sum(All.enh.TE),\n",
    "           \"Shared\"=Both.enh.TE[names(colors.TE)]/sum(Both.enh.TE),\n",
    "           \"Husk\"=husk.enh.TE[names(colors.TE)]/sum(husk.enh.TE), \n",
    "           \"V2-IST\"=V2.enh.TE[names(colors.TE)]/sum(V2.enh.TE))\n",
    "\n",
    "### Compute corresponding Odds ratio and p-values using fisher tests\n",
    "res.enrich.TE.counts <- NULL\n",
    "for(te in names(colors.TE)[!(names(colors.TE) %in% c(\"None\", \"Multiple\"))]){\n",
    "    real.tmp <- tapply(enh.TE$order, enh.TE$enh.pos, function(x){any(x == te)})\n",
    "    fh <- fisher.test(table(real.tmp, (tissue.enh[names(real.tmp)]==\"Husk\")))\n",
    "    fv <- fisher.test(table(real.tmp, (tissue.enh[names(real.tmp)]==\"V2-IST\")))\n",
    "    fb <- fisher.test(table(real.tmp, (tissue.enh[names(real.tmp)]==\"Both\")))\n",
    "    res.enrich.TE.counts <- rbind(res.enrich.TE.counts, \n",
    "                               c(fh$p.value, fh$estimate,\n",
    "                                 fv$p.value, fv$estimate,\n",
    "                                 fb$p.value, fb$estimate))\n",
    "}\n",
    "rownames(res.enrich.TE.counts) <- names(colors.TE)[!(names(colors.TE) %in% c(\"None\", \"Multiple\"))]\n",
    "colnames(res.enrich.TE.counts) <- c(\"p-value Husk\", \"OR Husk\", \"p-value V2-IST\", \"OR V2-IST\", \"p-value Shared\", \"OR Shared\")\n",
    "res.enrich.TE.counts\n",
    "\n",
    "### Plot the result as a barplot\n",
    "par(mar=c(4,7,7,2)+.1, las=1)\n",
    "barplot(prop, horiz=T, col=colors.TE, cex.axis=1.5, cex.names=1.8, cex.lab=1.8, xlab=\"Proportion of enhancers overlapping TEs\")\n",
    "par(\"xpd\"=TRUE)\n",
    "legend((par(\"usr\")[1]-(par(\"usr\")[2]-par(\"usr\")[1])/4), (par(\"usr\")[4]+(par(\"usr\")[4]-par(\"usr\")[3])/4), \n",
    "       legend=names(colors.TE), \n",
    "       fill=colors.TE,\n",
    "       bty=\"n\", ncol=3, cex=1.8)\n",
    "par(\"xpd\"=FALSE)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, we investigate the proportion of TFBS predicted in husk-specific enhancers that are overlapping at least one TE.  \n",
    "\n",
    "We find that husk-specific TFBS do not overlap more frequently than expected by chance any particular TE order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(paste0(ppath, \"tfbs_enhancers_TE_genes.Rdata\"))\n",
    "\n",
    "# Extract tfbs and TE information from table\n",
    "tfbs.TE = tfbs.TE.genes\n",
    "tfbs.TE[, geneID := NULL]\n",
    "tfbs.TE[, gene.tss := NULL]\n",
    "tfbs.TE <- unique(tfbs.TE)\n",
    "\n",
    "### Test for TE order enrichment among husk TFBS\n",
    "res.enrich.TFBS.husk.order=NULL\n",
    "tissue <- tfbs.TE$tissue\n",
    "for(te in c(\"TIR\", \"MITE\", \"LTR\", \"LINE\")){\n",
    "  real.tmp <- (tfbs.TE$order == te)\n",
    "  fh <- fisher.test(table(real.tmp, (tissue==\"Husk\")))\n",
    "  res.enrich.TFBS.husk.order=rbind(res.enrich.TFBS.husk.order, \n",
    "                             c(fh$p.value, fh$estimate))\n",
    "  }\n",
    "rownames(res.enrich.TFBS.husk.order) <- c(\"TIR\", \"MITE\", \"LTR\", \"LINE\")\n",
    "colnames(res.enrich.TFBS.husk.order) <- c(\"p-value Husk\", \"OR Husk\")\n",
    "\n",
    "res.enrich.TFBS.husk.order"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We perform the same analysis and find that TFBS predicted in V2-IST-specific enhancers preferentially overlap Terminal Inverted Repeat sequences (TIR). We then investigate if any specific superfamily of TIR is responsible for this enrichment, and find that they are enriched in TIR Mutator (DTM) elements."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Test for TE order enrichment among TFBS\n",
    "res.enrich.TFBS.V2.order=NULL\n",
    "tissue <- tfbs.TE$tissue\n",
    "for(te in c(\"TIR\", \"MITE\", \"LTR\", \"LINE\")){\n",
    "  real.tmp <- (tfbs.TE$order == te)\n",
    "  fh <- fisher.test(table(real.tmp, (tissue==\"V2-IST\")))\n",
    "  res.enrich.TFBS.V2.order=rbind(res.enrich.TFBS.V2.order, \n",
    "                             c(fh$p.value, fh$estimate))\n",
    "  }\n",
    "rownames(res.enrich.TFBS.V2.order) <- c(\"TIR\", \"MITE\", \"LTR\", \"LINE\")\n",
    "colnames(res.enrich.TFBS.V2.order) <- c(\"p-value V2-IST\", \"OR V2-IST\")\n",
    "\n",
    "res.enrich.TFBS.V2.order\n",
    "\n",
    "### Test for TE superfamily enrichment among TFBS\n",
    "res.enrich.TFBS.V2.sf <- NULL\n",
    "TE.order.sf <- unique(paste(tfbs.TE$order, tfbs.TE$superfamily, sep=\"_\"))\n",
    "tissue <- tfbs.TE$tissue\n",
    "for(te in c(\"TIR_DTA\", \"TIR_DTC\", \"TIR_DTM\")){\n",
    "  real.tmp <- (paste(tfbs.TE$order, tfbs.TE$superfamily, sep=\"_\")== te)\n",
    "  fh <- fisher.test(table(real.tmp, (tissue==\"V2-IST\")))\n",
    "  res.enrich.TFBS.V2.sf <- rbind(res.enrich.TFBS.V2.sf, \n",
    "                              c(fh$p.value, fh$estimate, \n",
    "                                fh$conf.int[1], fh$conf.int[2]))\n",
    "}\n",
    "rownames(res.enrich.TFBS.V2.sf) <- c(\"TIR_DTA\", \"TIR_DTC\", \"TIR_DTM\")\n",
    "colnames(res.enrich.TFBS.V2.sf) <- c(\"p-value V2-IST\", \"OR V2-IST\", \"CI Inf V2-IST\",  \"CI Sup V2-IST\")\n",
    "\n",
    "res.enrich.TFBS.V2.sf\n",
    "\n",
    "#plotting results for superfamilies\n",
    "res.enrich.TFBS.V2.sf <- data.frame(res.enrich.TFBS.V2.sf[3:1,])\n",
    "color <- c(rep(\"forestgreen\",3) )\n",
    "\n",
    "par(las=1, mar=c(4,7,0,0)+.1)\n",
    "plot(log2(res.enrich.TFBS.V2.sf$OR.V2.IST), c(1:nrow(res.enrich.TFBS.V2.sf)), \n",
    "     col=color, pch=18, cex=3, xlab=\"Odds ratio\", ylab=\"\", yaxt=\"n\", ylim=c(0.5,3.5),\n",
    "     xlim=c(min(log2(res.enrich.TFBS.V2.sf$CI.Inf.V2.IST)), max(log2(res.enrich.TFBS.V2.sf$CI.Sup.V2.IST))), \n",
    "     cex.axis=1.5, cex.lab=1.8)\n",
    "for(i in 1:3){\n",
    "  lines(c(log2(res.enrich.TFBS.V2.sf$CI.Inf.V2.IST[i]), log2(res.enrich.TFBS.V2.sf$CI.Sup.V2.IST[i])), c(i,i), \n",
    "      col=color[i], lwd=3)\n",
    "  lines(rep(log2(res.enrich.TFBS.V2.sf$CI.Inf.V2.IST[i]),2), c((i-0.1), (i+0.1)), \n",
    "        col=color[i], lwd=3)\n",
    "  lines(rep(log2(res.enrich.TFBS.V2.sf$CI.Sup.V2.IST[i]),2), c((i-0.1), (i+0.1)) , \n",
    "        col=color[i], lwd=3)\n",
    "}\n",
    "lines(c(0,0), c(-1,8), lwd=3)\n",
    "axis(side=2, labels=gsub(\"_\", \" \", rownames(res.enrich.TFBS.V2.sf)), at=1:3, cex.axis=1.8, cex.lab=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we look at the biological functions of the genes that are regulated by TFBS overlapping TIR Mutator elements. They are enriched in functions linked to cell growth (regulation of cell growth, cell wall polysaccharide biosynthetic process) and plant tissue differentiation (xylan biosynthetic process)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(paste0(ppath, \"tfbs_enhancers_TE_genes.Rdata\"))\n",
    "if(precomputed==0){\n",
    "    DTM.targets <- unique(sort(tfbs.TE.genes[order==\"TIR\" & superfamily==\"DTM\"]$geneID))\n",
    "    TIR_GO_res <- GOenrich(geneNames = geneNames, geneID2GO = geneID2GO,\n",
    "         list.topgenes = list(\"DTM\"=DTM.targets), all.genes = unique(unlist(V2_genes)), tag = \"TIR_targets\",  max.nb.genes = 100)\n",
    "} else {\n",
    "    load(paste0(ppath, \"GO_TIR_targets.Rdata\"))\n",
    "}\n",
    "tmp <- TIR_GO_res[[1]][[1]]$DTM\n",
    "tmp <- tmp[order(tmp$elim), ]\n",
    "head(tmp, n = 15)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# References\n",
    "\n",
    "1. <div class=\"csl-entry\">Spitz, F. &#38; Furlong, E. E. M. (2012). Core promoter Transcription factors: from enhancer binding to developmental control. <i>NATURE REVIEWS | GENETICS</i>, <b>13</b>:613. https://doi.org/10.1038/nrg3207 </div>\n",
    "\n",
    "2. <div class=\"csl-entry\">Weber, B., Zicola, J., Oka, R., Stam, M. (2016). Plant Enhancers: A Call for Discovery. <i>Trends in Plant Science</i>, <b>21</b>(11):974–987. https://doi.org/10.1016/J.TPLANTS.2016.07.013 </div>\n",
    "\n",
    "3. <div class=\"csl-entry\">Zhu, B., Zhang, W., Zhang, T., Liu, B., Jiang, J. (2015). Genome-Wide Prediction and Validation of Intergenic Enhancers in Arabidopsis Using Open Chromatin Signatures. <i>The Plant Cell</i>, <b>27</b>(9):2415–2426. https://doi.org/10.1105/TPC.15.00537 </div>\n",
    "\n",
    "4. <div class=\"csl-entry\">Ricci, W. A., Lu, Z., Ji, L., Marand, A. P., Ethridge, C. L., Murphy, N. G., Noshay, J. M., Galli, M., Mejía-Guerra, M. K., Colomé-Tatché, M., Johannes, F., Rowley, M. J., Corces, V. G., Zhai, J., Scanlon, M. J., Buckler, E. S., Gallavotti, A., Springer, N. M., Schmitz, R. J., Zhang, X. (2019). Widespread long-range cis-regulatory elements in the maize genome. <i>Nature Plants</i>, <b>5</b>(12):1237–1249. https://doi.org/10.1038/S41477-019-0547-0 </div>\n",
    "\n",
    "5. <div class=\"csl-entry\">Lu, Z., Marand, A. P., Ricci, W. A., Ethridge, C. L., Zhang, X., Schmitz, R. J. (2019). The prevalence, evolution and chromatin signatures of plant regulatory elements. <i>Nature Plants</i>, <b>5</b>(12):1250–1259. https://doi.org/10.1038/S41477-019-0548-Z </div>\n",
    "\n",
    "6. <div class=\"csl-entry\">Li, E., Liu, H., Huang, L., Zhang, X., Dong, X., Song, W., Zhao, H., Lai, J. (2019). Long-range interactions between proximal and distal regulatory regions in maize. <i>Nature Communications</i>, <b>10</b>(1). https://doi.org/10.1038/S41467-019-10603-4 </div>\n",
    "\n",
    "7. <div class=\"csl-entry\">Oka, R., Zicola, J., Weber, B., Anderson, S. N., Hodgman, C., Gent, J. I., Wesselink, J. J., Springer, N. M., Hoefsloot, H. C. J., Turck, F.,  Stam, M. (2017). Genome-wide mapping of transcriptional enhancer candidates using DNA and chromatin features in maize. <i>Genome Biology</i>, <b>18</b>(1). https://doi.org/10.1186/S13059-017-1273-4 </div>\n",
    "\n",
    "8.  <div class=\"csl-entry\">Fagny, M., Kuijjer, M. L., Stam, M., Joets, J., Turc, O., Rozière, J., Pateyron, S., Venon, A., Vitte, C. (2021). Identification of Key Tissue-Specific, Biological Processes by Integrating Enhancer Information in Maize Gene Regulatory Networks. <i>Frontiers in Genetics</i>, <b>11</b>:1703. https://doi.org/10.3389/FGENE.2020.606285/BIBTEX </div>\n",
    "\n",
    "9.  <div class=\"csl-entry\">Grant, C. E., Bailey, T. L., Noble, W. S.(2011). FIMO: scanning for occurrences of a given motif. <i>Bioinformatics</i>, <b>27</b>(7):1017–1018. https://doi.org/10.1093/bioinformatics/btr064 </div>"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Format de la Cellule Texte Brut",
  "jupytext": {
   "cell_metadata_filter": "-all",
   "main_language": "R",
   "notebook_metadata_filter": "-all"
  },
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
